docs/user-guide/optimization.md

```diff
@@ -0,0 +1,109 @@
+
+Note that with optimization turned on, the array `b` is no longer written as an intermediate output since it will be computed in the same tasks that compute array `c`. The overall number of tasks is reduced from 10 to 5, and the intermediate data (total nbytes) is reduced too.
+
+## Advanced optimization
+
+Cubed supports more powerful optimizations, but these are not (currently) enabled by default. This section explains what they are and how you can try them out.
+
+Map fusion is limited to the cases where there is a single preceding input with the same number of tasks. However, many computations have a branching structure, and the default optimization doesn't help in such cases. Here is an example, where even though optimization is enabled, it doesn't have any effect.
+
+```python
+import cubed.array_api as xp
+
+a = xp.ones((3, 3), chunks=(2, 2))
+b = xp.ones((3, 3), chunks=(2, 2))
+c = xp.ones((3, 3), chunks=(2, 2))
+d = xp.add(b, c)
+e = xp.add(a, d)
+
+e.visualize()
+```
+
+![Multiple inputs unoptimized](../images/optimization_multiple_inputs_unoptimized.svg)
+
+To overcome this limitation we can use the `optimize_function` argument, which allows us to specify the function that performs optimization. Here we use the `multiple_inputs_optimize_dag` function that can handle multiple inputs.
+
+```python
+from cubed.core.optimization import multiple_inputs_optimize_dag
+
+e.visualize(optimize_function=multiple_inputs_optimize_dag)
+```
+
+![Multiple inputs unoptimized](../images/optimization_multiple_inputs.svg)
+
+Notice how the array `d` is fused away.
+
+The `multiple_inputs_optimize_dag` function has a couple of parameters to control how it works:
+
+`max_total_source_arrays` specifies the maximum number of source arrays that are allowed in the fused operation, in order to limit the number of reads that an individual task must perform.
+
+The default is 4. In the previous example above the fused operation has three source arrays (`a`, `b`, and `c`), which is below the maximum default allowed. On the other hand, a computation with a higher "fan-in" that exceeds the maximum will not be fused, or operations will be fused in stages.
+
+To change this setting, you can use Python's `functools.partial` as follows:
+
+```python
+opt_fn = partial(multiple_inputs_optimize_dag, max_total_source_arrays=8)
+visualize(optimize_function=opt_fn)
+```
+
+`max_total_num_input_blocks` specifies the maximum number of input blocks (chunks) that are allowed in the fused operation.
+
+Again, this is to limit the number of reads that an individual task must perform. The default is `None`, which means that operations are fused only if they have the same number of tasks. If set to an integer, then this limitation is removed, and tasks with a different number of tasks will be fused - as long as the total number of input blocks does not exceed the maximum. This setting is useful for reductions, and can be set using `functools.partial`:
```

Comment:
My understanding is that both the `max_total_source_arrays` and `max_total_num_input_blocks` are related to the "optimal fan-in" parameter described in #331. I therefore think this explanation would be better motivated if it started with a description of that general trade-off first. 

I would have structured it like this:
- Description of the trade-off detailed in #331
- Therefore an optimal fan-in parameter exists (unique for any particular reduction?)
- Cubed makes a default global choice for you, but you may want to override it
- In which scenarios are you likely to want to override this default?
- Here's how you actually use the API to change the default...